File: Top K Frequent Elements in Array -GFG/Top K Frequent Elements in Array .cpp
class Solution {
  public:
       vector<int> topK(vector<int>& nums, int k)
    {

        int N = nums.size();

        unordered_map<int, int> mp;
        for (int i = 0; i < N; i++)
            mp[nums[i]]++;

        priority_queue<pair<int, int> > pq;

        for (auto x : mp) {//N
            pq.push({ x.second, x.first });  //logN
        }
    //     while (!pq.empty()) {
    // cout << pq.top().first << " " << pq.top().second << endl;
    // pq.pop();
    // }
    // freq : ele
    // 3 2
    // 3 1
    // 1 3
      
        vector<int> ans;

        for (int i = 1; i <= k; i++) {
            ans.push_back(pq.top().second);
            pq.pop();
        }

        return ans;
    }
};

File: Top K Frequent Elements in Array -GFG/Top K Frequent Elements in Array.java

class Solution {
    public int[] topK(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) { //N
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        PriorityQueue<Map.Entry<Integer, Integer>> pq = new PriorityQueue<>( //min heap by default
        //and we cannot use max heap since we wan top k largest
            (a, b) -> {//CUSTOM COMPARATOR
                if (!a.getValue().equals(b.getValue())) {
                    return a.getValue() - b.getValue(); //  frequency in asc
                } else {
                    return a.getKey() - b.getKey(); // if freq same, sort by key in asc
                }
            }
        );

        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {//N
            pq.add(entry);
            if(pq.size()>k)//logK
            pq.poll();
         
        }
    //  while(!pq.isEmpty()){
    //  System.out.println(pq.poll());
    //  }
        
        int[] ans = new int[k];
        int idx=k-1;
        while(!pq.isEmpty())
        {
            ans[idx--]=pq.poll().getKey();
        }
      return ans;

   
